pyvrp/cpp/SubPopulation_bindings.cpp

```diff
@@ -0,0 +1,67 @@
+#include "SubPopulation.h"
+
+#include <pybind11/functional.h>
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+
+namespace py = pybind11;
+
+PYBIND11_MODULE(SubPopulation, m)
+{
+    py::class_<PopulationParams>(m, "PopulationParams")
+        .def(py::init<size_t, size_t, size_t, size_t, double, double>(),
+             py::arg("min_pop_size") = 25,
+             py::arg("generation_size") = 40,
+             py::arg("nb_elite") = 4,
+             py::arg("nb_close") = 5,
+             py::arg("lb_diversity") = 0.1,
+             py::arg("ub_diversity") = 0.5)
+        .def_readwrite("min_pop_size", &PopulationParams::minPopSize)
+        .def_readwrite("generation_size", &PopulationParams::generationSize)
+        .def_property_readonly("max_pop_size", &PopulationParams::maxPopSize)
+        .def_readwrite("nb_elite", &PopulationParams::nbElite)
+        .def_readwrite("nb_close", &PopulationParams::nbClose)
+        .def_readwrite("lb_diversity", &PopulationParams::lbDiversity)
+        .def_readwrite("ub_diversity", &PopulationParams::ubDiversity);
+
+    py::class_<SubPopulation::Item>(m, "SubPopulationItem")
```

Comment:
Can we not expose the SubPopulationItem through the bindings? I think this is an implementation detail. I suggest `SubPopulation::__getitem__` returns an Individual and `SubPopulation::__iter__` iterates over Individuals. Preferably also on the c++ code but at least on the bindings. To facility the internal tests and sampling the binary tournament, you could make a `get_diversity(idx)` method (not pretty to use idx but same `SubPopulation::avg_distance_closest`. For the test, you can expose an `_get_proximity` method if you really need to. This keeps implementation details abstracted from the Python interface.

Reply:
I think the entirety of `SubPopulation` is an implementation detail. The only reason it might not be is that we have `feasible_subpopulation` and `infeasible_subpopulation` properties on the `Population`. But those feel like a step in the wrong direction (why does a user need to know the Population has subpopulations by feasibility?), and should not be part of the public interface.

> To facility the internal tests and sampling the binary tournament, you could make a `get_diversity(idx)` method (not pretty to use idx but same `SubPopulation::avg_distance_closest`.

I think this is a mistake, for the same reason having `avgDistanceClosest` with an index was a mistake. That method should be a property of the item, not a standalone thing on `SubPopulation`.